### PR TITLE
Use init_once around library initialization

### PR DESCRIPTION
**Pull request by lmctv**
Making for a race free version of #187 @asn-the-goblin-slayer,  @JackWink

### CODE REVIEW

File: src/nacl/bindings/sodium_core.py [hunk 1]
```diff
@@ -13,14 +13,18 @@
...
 from __future__ import absolute_import, division, print_function
 
-from nacl._sodium import lib
+from nacl._sodium import ffi, lib
 from nacl.exceptions import CryptoError
 
 
-def sodium_init():
+def _sodium_init():
     """
```
Comment (dstufft) [inline src/nacl/bindings/sodium_core.py]: Docstring should be on public function.

Comment (lmctv) [inline src/nacl/bindings/sodium_core.py]: Sorry! Updated PR in flight.

**Comment by asn-the-goblin-slayer** [comment]
Thx for fixing this!

**Comment by JackWink** [comment]
I can't remember if PyNaCl allows you to bring your own libsodium or if it uses a bundled version, but libsodium has made `sodium_init` thread safe starting in version 1.0.11. 

This seems like an alright fix for older versions of libsodium, though one possible issue is that you are no longer able to call `sodium_init` more than once. That said, I'm not sure if there is a legitimate case where you'd want to call `sodium_init` multiple times.

**Comment by lmctv** [comment]
In as much as I can see, sodium_init itself is written to avoid repeated initializations.  The ffi.init_once call would still avoid useless crossing of the c/python boundary.

From what I see the boundled version is still 1.0.8 which predates the 1.0.11 thread safety enhancements.

Thank you very much.

Il 19 agosto 2016 21:58:03 CEST, Jack Wink <notifications@github.com> ha scritto:
>I can't remember if PyNaCl allows you to bring your own libsodium or if
>it uses a bundled version, but libsodium has made `sodium_init` thread
>safe starting in version 1.0.11. 
>
>This seems like an alright fix for older versions of libsodium, though
>one possible issue is that you are no longer able to call `sodium_init`
>more than once. That said, I'm not sure if there is a legitimate case
>where you'd want to call `sodium_init` multiple times.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/pyca/pynacl/pull/199#issuecomment-241119519

+-------------------------+----------------------------------------------+
|   Lorenzo M.  Catucci   | Centro di Calcolo e Documentazione           |
| catucci@ccd.uniroma2.it | Università degli Studi di Roma "Tor Vergata" |
|                         | Via O. Raimondo 18 ** I-00173 ROMA  ** ITALY |
|  Tel. +39 06 7259 2255  | Fax. +39 06 7259 2125                        |
+-------------------------+----------------------------------------------+

**Comment by JackWink** [comment]
Agreed, I think this is an acceptable fix. It looks like PyNaCl supports bringing your own libsodium, so even if the bundled version was updated, we would still want to handle thread safety from the python side for older system distributions.

**Comment by lmctv** [comment]
@warner  would you mind taking a look at this PR? 633ece35ac0df3e3f2deb656b1f121d97af3e5d3 is needed also for all of the other ones, since they use ffi.memmove, which has been implemented in 1.3,

**Pull request merged**